python/instrumentation/openinference-instrumentation-bedrock/src/openinference/instrumentation/bedrock/__init__.py

```diff
@@ -158,7 +164,75 @@ def instrumented_response(*args: Any, **kwargs: Any) -> Dict[str, Any]:
+                    _set_span_attribute(span, SpanAttributes.LLM_MODEL_NAME, model_id)
+
+                if request_messages := kwargs.get("messages"):
+                    # Extract the most recent request message in message history list
+                    current_request_message = request_messages[-1]
+
+                    if request_content := current_request_message.pop("content", ""):
+                        # Currently only supports single text-based data
+                        # Future implementation can be extended to handle different media types
+                        if prompt := request_content[-1].get("text"):
+                            _set_span_attribute(span, SpanAttributes.INPUT_VALUE, prompt)
+
+                    invocation_parameters = safe_json_dumps(current_request_message)
+                    _set_span_attribute(
+                        span, SpanAttributes.LLM_INVOCATION_PARAMETERS, invocation_parameters
+                    )
+
+                    # Add content back after setting invocation parameters
+                    if request_content:
+                        current_request_message["content"] = request_content
+
+                response = wrapped_client._unwrapped_converse(*args, **kwargs)
+                if (response_content :=
+                        response.get("output", {}).get("message", {}).get("content", [])):
+                    current_content = response_content[-1]
+                    # Currently only supports single text-based data
+                    if response_text := current_content.get("text"):
+                        _set_span_attribute(span, SpanAttributes.OUTPUT_VALUE, response_text)
+
+                if usage := response.get("usage"):
+                    if input_token_count := usage.get("inputTokens"):
+                        _set_span_attribute(
+                            span, SpanAttributes.LLM_TOKEN_COUNT_PROMPT, input_token_count
+                        )
+                    if response_token_count := usage.get("outputTokens"):
+                        _set_span_attribute(
+                            span,
+                            SpanAttributes.LLM_TOKEN_COUNT_COMPLETION,
+                            response_token_count,
+                        )
+                    if total_token_count := usage.get("totalTokens"):
+                        _set_span_attribute(
+                            span, SpanAttributes.LLM_TOKEN_COUNT_TOTAL, total_token_count
+                        )
+
+                span.set_attributes(dict(get_attributes_from_context()))
+            return response  # type: ignore
+
+        def dispatcher(method_name):
+            if method_name == "invoke_model":
+                return instrumented_response_invoke
+            elif method_name == "converse":
+                return instrumented_response_converse
+            else:
+                raise ValueError(f"Unsupported method: {method_name}")
+
+        return dispatcher
```

Comment:
when does this get called? is this going to cause an issue if a method that isn't invoke or converse gets called on the client? I don't have a great understanding of how this works but would be very carefl about throwing in here since something that is supposed to be patching behind the scenes will now be breaking application logic potentially

Reply:
_model_invocation_wrapper is only ever called/assigned to inovke_model and converse so no other method for the client should touch this wrapper. I also thought of using a partial function instead of this dispatcher function. Would love to get some feedback on what might be safest.

Reply:
^ agree we don't need a dispatcher, we can define two separate wrapper functions since we explicitly wrap both converse and invoke earlier in the file